### PR TITLE
Feature: Drawing Annotations Undo Redo

### PR DESCRIPTION
**Pull request by MinhHNguyen**
Contains changes from #267 that need to be merged in first.

- Added an event observer function in annotator.bindModeListeners that will be moved to a new controller class in upcoming changes. The controller will use an RTree to contain each DrawingThread and its axis aligned bounding box. It might also help clean up some logic in annotator.
- Added undo and redo functionality for drawings that have not yet been committed

### CODE REVIEW

**Comment by MinhHNguyen** [comment]
Unit tests are in

File: src/lib/annotations/Annotator.js [hunk 1]
```diff
@@ -677,23 +690,53 @@ class Annotator extends EventEmitter {
...
-            handlers.push({
-                type: 'mouseup',
-                func: annotatorUtil.eventToLocationHandler(locationFunction, drawingThread.handleStop),
-                eventObj: this.annotatedElement
+            const undoButtonEl = this.previewUI.getAnnotateButton(SELECTOR_ANNOTATION_BUTTON_DRAW_UNDO);
+            const redoButtonEl = this.previewUI.getAnnotateButton(SELECTOR_ANNOTATION_BUTTON_DRAW_REDO);
+
+            // NOTE (@minhnguyen): Move this logic to a new controller class
+            const that = this;
+            drawingThread.addListener('annotationevent', (data = {}) => {
```
Comment (pramodsum) [inline src/lib/annotations/Annotator.js]: move this into a separate method + tests

File: src/lib/annotations/__tests__/Annotator-test.js [hunk 2]
```diff
@@ -560,7 +536,7 @@ describe('lib/annotations/Annotator', () => {
                     'click',
                     sinon.match.func
                 );
-                expect(annotator.annotationModeHandlers.length).equals(4);
+                expect(annotator.annotationModeHandlers.length).equals(6);
```
Comment (pramodsum) [inline src/lib/annotations/__tests__/Annotator-test.js]: Add tests for when only some of the buttons exist i.e postButtonEl, undoButton...etc

File: src/lib/annotations/doc/DocAnnotator.js [hunk 3]
```diff
@@ -490,9 +490,11 @@ class DocAnnotator extends Annotator {
         // We need to redraw highlights on the page if a thread was deleted
         // since deleting 'cuts' out the highlight, which may have been
         // overlapping with another
-        thread.addListener('threaddeleted', () => {
-            this.showHighlightsOnPage(thread.location.page);
-        });
+        if (annotatorUtil.isHighlightAnnotation(thread.type)) {
```
Comment (pramodsum) [inline src/lib/annotations/doc/DocAnnotator.js]:
```
if (!annotatorUtil.isHighlightAnnotation(thread.type)) {
    return;
}

thread.addListener(.....
```

File: src/lib/annotations/doc/DocDrawingThread.js [hunk 4]
```diff
@@ -116,16 +114,23 @@ class DocDrawingThread extends DrawingThread {
      * @return {void}
      */
     saveAnnotation(type, text) {
-        super.saveAnnotation(type, text);
+        this.emit('threadcleanup');
```
Comment (pramodsum) [inline src/lib/annotations/doc/DocDrawingThread.js]: What's the reason for deleting the thread on save?

File: src/lib/annotations/drawing/DrawingContainer.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    /**
+     * Retrieve a JSON blob containing the number of undo and redo in each stack.
+     *
+     * @return {AvailableItemData} The number of undo and redo items available.
+     */
+    getNumberOfItems() {
+        return {
+            undo: this.undoStack.length,
```
Comment (pramodsum) [inline src/lib/annotations/drawing/DrawingContainer.js]: maybe rename to undoCount and redoCount for clarity?

File: src/lib/annotations/doc/DocDrawingThread.js (hunk 4 above)
Comment (MinhHNguyen) [inline src/lib/annotations/doc/DocDrawingThread.js]: Clean-up removes the listeners, 'threaddeleted' deletes the thread from the thread map.

File: src/lib/annotations/Annotator.js (hunk 1 above)
Comment (MinhHNguyen) [inline src/lib/annotations/Annotator.js]: This will be done in a separate PR. I don't think it makes sense to separate and test it at the moment when it will be moved in a coming class change.

File: src/lib/annotations/doc/DocDrawingThread.js (hunk 4 above)
Comment (MinhHNguyen) [inline src/lib/annotations/doc/DocDrawingThread.js]: Realized that bindings might be made to bind to 'threaddeleted' so I'll add a custom 'annotationevent' of type 'drawingcommit' to removeAllListeners('annotationevent'). We should be able to clean this up in the future once some logic is pulled out of Annotator.
